flink-test-utils-parent/flink-migration-test-utils/src/main/java/org/apache/flink/test/migration/MigrationTestsSnapshotGenerator.java

```diff
@@ -0,0 +1,152 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * This class aims to generate the snapshots for all the state migration tests. A migration tests
+ * should implement {@link MigrationTest} interface and its name should match {@code
+ * *(Test|ITCase)(.java|.scala)}. For scala tests, we also require the class name is the same with
+ * the containing filename.
+ */
+public class MigrationTestsSnapshotGenerator {
+
+    private static final Logger LOG =
+            LoggerFactory.getLogger(MigrationTestsSnapshotGenerator.class);
+
+    private static final String[] DEFAULT_PATH_PREFIXES =
+            new String[] {"src/test/java", "src/test/scala"};
+
+    private static final Pattern VERSION_PATTERN = Pattern.compile("" + "v?([0-9]+)[._]([0-9]+)");
+
+    private static final String CLASS_NAME_GROUP = "className";
+    private static final Pattern CLASS_NAME_PATTERN =
+            Pattern.compile("(?<" + CLASS_NAME_GROUP + ">[a-zA-Z0-9]*(Test|ITCase))(.java|.scala)");
+
+    public static void main(String[] args) {
+        try {
+            if (args.length < 2) {
+                throw new IllegalArgumentException(
+                        "Usage: java MigrationTestsSnapshotGenerator [project root path] [version]");
+            }
+
+            File rootDirectory = new File(args[0]);
+            if (!rootDirectory.exists() || !rootDirectory.isDirectory()) {
+                throw new FileNotFoundException(
+                        rootDirectory + " does not exist or is not a directory.");
+            }
+
+            String versionName = args[1];
+            Matcher versionMatcher = VERSION_PATTERN.matcher(versionName);
+            if (!versionMatcher.matches()) {
+                throw new IllegalArgumentException(
+                        "Version "
+                                + versionName
+                                + "could not be parsed, "
+                                + "please specify the version with format like 1.17, 1_17, v1_17, v1.17");
+            }
+            String normalizedVersionName =
+                    "v" + versionMatcher.group(1) + "_" + versionMatcher.group(2);
+            FlinkVersion version = FlinkVersion.valueOf(normalizedVersionName);
+
+            LOG.info("Start generating for module {} and version {}", rootDirectory, version);
+            SnapshotGeneratorExecutor executor = new SnapshotGeneratorExecutor();
+            List<Class<?>> migrationTests = findMigrationTests(rootDirectory.getAbsolutePath());
+            for (Class<?> migrationTestClass : migrationTests) {
+                LOG.info("Start generating for {}", migrationTestClass.getName());
+                executor.executeGenerate(migrationTestClass, version);
+                LOG.info("Finish generating for {}", migrationTestClass.getName());
+            }
+
+            // Avoids leaking threads blocks the process.
+            System.exit(0);
```

Comment:
I have a double check on the reason of the leaking threads, most of them comes from the static fields used in some tests, like https://github.com/apache/flink/blob/009a9e5ce7dc967cb3a1e7f5d997835707e05ccd/flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionGraphPartitionReleaseTest.java#L62. 

Since in the current implementation we load and initialization every class, these static fields will be initialized. I modified the logic here to load without initializing the unrelated classes.

Reply:
I created FLINK-30944 to cover the thread leaking in `ExecutionGraphPartitionReleaseTest`

Reply:
The code change is roughly at org.apache.flink.test.migration.MigrationTestsSnapshotGenerator:L236:

```java
Class<?> clazz =
        Class.forName(
                className,
                false,
                MigrationTestsSnapshotGenerator.class.getClassLoader());
if (MigrationTest.class.isAssignableFrom(clazz)) {
    result.add(Class.forName(className));
}
```

Here it first load the class without initializing it to check if it is a migration test, and only do the initializing if it is indeed a migration test.

The `${generated.classes}` is used to pass the `classes` parameter to the `MigrationTestsSnapshotGenerator`. I thought this parameter might be useful if some developers add a new migration test, and want to generate snapshots for this specific test.